test/sparse/sparse.jl

```diff
@@ -1,469 +1,518 @@
+            @test_throws DimensionMismatch permute!(A[:, 1:(m - 1)], A, p, q)
+            @test_throws ArgumentError permute!((Y = copy(X); resize!(Y.rowval, nnz(A) - 1); Y), A, p, q)
+            @test_throws ArgumentError permute!((Y = copy(X); resize!(Y.nzval, nnz(A) - 1); Y), A, p, q)
+            # Test common error checking of permute[!] methods / source-workmat compat
+            @test_throws DimensionMismatch permute!(X, A, p, q, C[1:(m - 1), :])
+            @test_throws DimensionMismatch permute!(X, A, p, q, C[:, 1:(m - 1)])
+            @test_throws ArgumentError permute!(X, A, p, q, (D = copy(C); resize!(D.rowval, nnz(A) - 1); D))
+            @test_throws ArgumentError permute!(X, A, p, q, (D = copy(C); resize!(D.nzval, nnz(A) - 1); D))
+            # Test common error checking of permute[!] methods / source-workcolptr compat
+            @test_throws DimensionMismatch permute!(A, p, q, C, Vector{eltype(A.rowval)}(length(A.colptr) - 1))
+            # Test common error checking of permute[!] methods / permutation validity
+            @test_throws ArgumentError permute!(A, (r = copy(p); r[2] = r[1]; r), q)
+            @test_throws ArgumentError permute!(A, (r = copy(p); r[2] = m + 1; r), q)
+            @test_throws ArgumentError permute!(A, p, (r = copy(q); r[2] = r[1]; r))
+            @test_throws ArgumentError permute!(A, p, (r = copy(q); r[2] = n + 1; r))
+        end
+        @testset "overall functionality of [c]transpose[!] and permute[!]" begin
+            for (m, n) in ((smalldim, smalldim), (smalldim, largedim), (largedim, smalldim))
+                A = sprand(m, n, nzprob)
+                At = transpose(A)
+                # transpose[!]
+                fullAt = transpose(Array(A))
+                @test transpose(A) == fullAt
+                @test transpose!(similar(At), A) == fullAt
+                # ctranspose[!]
+                C = A + im*A/2
+                fullCh = ctranspose(Array(C))
+                @test ctranspose(C) == fullCh
+                @test ctranspose!(similar(sparse(fullCh)), C) == fullCh
+                # permute[!]
+                p = randperm(m)
+                q = randperm(n)
+                fullPAQ = Array(A)[p,q]
+                @test permute(A, p, q) == sparse(Array(A[p,q]))
+                @test permute!(similar(A), A, p, q) == fullPAQ
+                @test permute!(similar(A), A, p, q, similar(At)) == fullPAQ
+                @test permute!(copy(A), p, q) == fullPAQ
+                @test permute!(copy(A), p, q, similar(At)) == fullPAQ
+                @test permute!(copy(A), p, q, similar(At), similar(A.colptr)) == fullPAQ
+            end
+        end
     end
 end
 
-# transpose of SubArrays
-A = view(sprandn(10, 10, 0.3), 1:4, 1:4)
-@test  transpose(Array(A)) == Array(transpose(A))
-@test ctranspose(Array(A)) == Array(ctranspose(A))
-
-# exp
-A = sprandn(5,5,0.2)
-@test e.^A ≈ e.^Array(A)
-
-# reductions
-pA = sparse(rand(3, 7))
+@testset "Transpose of subarrays" begin
```

Comment:
Perhaps "Transpose" should be "transpose", referring to the (lowercase) function name? Perhaps "subarrays" should be "SubArrays", referring to the (CamelCase) type name?